PEP 730: Adding iOS as a supported platform (#3473)
python/peps: peps/pep-0730.rst

CI resources
------------

Github Actions is able to host iOS simulators on their macOS machines, and the iOS simulator can be controlled by scripting environments. The free tier currently only provides x86_64 macOS machines; however ARM64 runners have recently become available on paid plans [5]_.

If Github Actions resources are insufficient or not viable for cost reasons, Anaconda has offered to provide resources to support CI requirements.

Packaging
---------

iOS will not provide a "universal" wheel format. Instead, wheels will be provided on a for each ABI-arch combination. At present, no binary merging is required. There is only 1 on-device architecture; and simulator binaries are not considered to be distributable artefacts, so only 1 architecture is needed to build a simulator.

iOS wheels will use tags:

* ``iOS_12_0_iphoneos_arm64``
* ``iOS_12_0_iphonesimulator_arm64``
* ``iOS_12_0_iphonesimulator_x86_64``

In these tags, "12.0" is the minimum supported iOS version. The choice of minimum supported iOS version is a decision of whoever compiles CPython for iOS. At time of writing, iOS 12.0 exposes most significant iOS features, while reaching near 100% of devices.

> These wheels can include binaries modules in-situ (i.e., co-located with the Python source, in the same way as wheels for a desktop platform); however, they will need to be post-processed as part of inin the samewill need to be post-processed into the "Frameworks" location required for distribution. This can be automated with an Xcode build step.

Reviewer: ```suggestion
These wheels can include binary modules in-situ (i.e., co-located with the Python source, in the same way as wheels for a desktop platform); however, they will need to be post-processed as binary modules need to be moved into the "Frameworks" location for distribution. This can be automated with an Xcode build step.
```
Not sure about the second sentence but the existing wording was broken

> PEP11 Update

Reviewer: ```suggestion
PEP 11 Update
```

------------

:pep:`11` will be updated to include the 3 iOS ABIs:

* ``aarch64-apple-ios``
* ``aarch64-apple-ios-simulator``
* ``x86_64-apple-ios-simulator``

Ned Deily will serve as the initial Core Team contact for these ABIs.

Backwards Compatibility
=======================

Adding a new platform does not introduce any backwards compatibility concerns to CPython
itself.

There may be some backwards compatibility implications on the projects that have
historically provided CPython support (i.e., BeeWare and Kivy) if the final form of any
CPython patches don't align with the patches they have historically used.

Although not strictly a backwards compatibility issue, there *is* a platform adoption
consideration. Although CPython itself may support iOS, if it is unclear how to produce

> iOS-compatibile wheels, and prominent libraries like cryptography, Pillow, and numpy

Reviewer: ```suggestion
iOS-compatible wheels, and prominent libraries like cryptography, Pillow, and numpy
```

don't provide iOS wheels, the ability of the community to adopt Python on iOS will be
limited. Therefore, it will be necessary to clearly document how projects can add iOS

> builds to their CI and release tooling.

Author: `cibuildwheel` is definitely worth mentioning here. I'll add a note to that effect.

Security Implications
=====================

Adding iOS as a new platform does not add any security implications.

How to Teach This
=================

> The education needs related to this PEP mostly relate to how end-users can add iOS support to their own Xcode projects. This can be accomplished with documentation and tutorials on that process. The need for this documentation will increase if/when support raises from Tier 3 to Tier 2 or 1; however, this transition should also be accompanied with simplified deployment artefacts (such as a Cocoapod or Swift ) that are integrated with Xcode development.

Reviewer: Missing a word after "Swift"?

* ``platform.node()`` - the user-provided name of the device, as returned by the
  ``[[UIDevice currentDevice] systemName]`` system call. (e.g., ``Janes-iPhone``). For
  simulated devices, this will be the name of the development computer running the
  simulator.

* ``platform.release()`` - the iOS version number, as a string (e.g., "16.6.1")

* ``platform.machine()`` - The device model returned by ``[[UIDevice currentDevice]
  model]`` (e.g., ``iPhone13,2``); or ``iPhoneSimulator`` for simulated devices.

All other values will be as returned by ``os.uname()``.

``os``
''''''

``os.uname()`` will return the raw result of a POSIX ``uname()`` call. This will result
in the following values:

* ``sysname`` - ``Darwin``

* ``release`` - The Darwin kernel version (e.g., ``"22.6.0"``)

``sysconfig``
'''''''''''''

The ``sysconfig`` module will use the minimum iOS version as part of ``sysconfig.get_platform()`` identifier (e.g., ``iOS-12.0-iphoneos-arm64``). The ``sysconfigdata_name`` and Config makefile will follow the same patterns as existing platforms (using ``sys.platform``, ``sys.implementation._multiarch`` etc) to construct identifiers.

Subprocess support
------------------

> iOS will leverage the pattern for disabling subprocesses established by WASI/Emscripten. The ``subprocess`` module will raise an exception if an attempt is made to start a subprocess ``os.fork`` and ``os.spawn`` calls will raise an ``OSError``.

Reviewer 2: ```suggestion
iOS will leverage the pattern for disabling subprocesses established by WASI/Emscripten. The ``subprocess`` module will raise an exception if an attempt is made to start a subprocess, and ``os.fork`` and ``os.spawn`` calls will raise an ``OSError``.
```

Dynamic module loading
----------------------

To accommodate iOS dynamic loading, the ``importlib`` bootstrap will be extended to add a metapath finder that can convert a request for a Python binary module into a Framework location. This finder will only be installed if ``sys.platform == "ios"``.

This finder will convert a Python module name (e.g., ``foo.bar._whiz``) into a unique Framework name by replacing the dots with underscores (i.e., ``foo_bar__whiz.framework``). A framework is a directory; the finder will look for ``_whiz.dylib`` in that directory.

CI resources
------------

GitHub Actions is able to host iOS simulators on their macOS machines, and the iOS simulator can be controlled by scripting environments. The free tier currently only provides x86_64 macOS machines; however ARM64 runners have recently become available on paid plans [5]_.

If GitHub Actions resources are insufficient or not viable for cost reasons, Anaconda has offered to provide resources to support CI requirements.

Packaging
---------

> iOS will not provide a "universal" wheel format. Instead, wheels will be provided for each ABI-arch combination. At present, no binary merging is required. There is only 1 on-device architecture; and simulator binaries are not considered to be distributable artefacts, so only 1 architecture is needed to build a simulator.

Reviewer 3: ```suggestion
iOS will not provide a "universal" wheel format. Instead, wheels will be provided for each ABI-arch combination. At present, no binary merging is required. There is only one on-device architecture; and simulator binaries are not considered to be distributable artefacts, so only one architecture is needed to build a simulator.
```

iOS wheels will use tags:

* ``iOS_12_0_iphoneos_arm64``
* ``iOS_12_0_iphonesimulator_arm64``
* ``iOS_12_0_iphonesimulator_x86_64``

In these tags, "12.0" is the minimum supported iOS version. The choice of minimum supported iOS version is a decision of whoever compiles CPython for iOS. At time of writing, iOS 12.0 exposes most significant iOS features, while reaching near 100% of devices.

These wheels can include binary modules in-situ (i.e., co-located with the Python source, in the same way as wheels for a desktop platform); however, they will need to be post-processed as binary modules need to be moved into the "Frameworks" location for distribution. This can be automated with an Xcode build step.

PEP 11 Update
-------------

> :pep:`11` will be updated to include the 3 iOS ABIs:

Reviewer 3: ```suggestion
:pep:`11` will be updated to include the three iOS ABIs:
```

* ``aarch64-apple-ios``
* ``aarch64-apple-ios-simulator``
* ``x86_64-apple-ios-simulator``

Ned Deily will serve as the initial Core Team contact for these ABIs.

Backwards Compatibility
=======================

Adding a new platform does not introduce any backwards compatibility concerns to CPython
itself.

There may be some backwards compatibility implications on the projects that have
historically provided CPython support (i.e., BeeWare and Kivy) if the final form of any
CPython patches don't align with the patches they have historically used.

Although not strictly a backwards compatibility issue, there *is* a platform adoption
consideration. Although CPython itself may support iOS, if it is unclear how to produce

> iOS-compatible wheels, and prominent libraries like cryptography, Pillow, and numpy

Reviewer 3: ```suggestion
iOS-compatible wheels, and prominent libraries like cryptography, Pillow, and NumPy
```

Reference Implementation
========================

The BeeWare Python-Apple-support [8]_ repository contains a reference patch and build tooling to compile a distributable artefact.

Briefcase [9]_ provides a reference implementation of code to execute test suites on iOS simulators. The Toga Testbed [10]_ is an example of a test suite that is executed on the iOS simulator using Github Actions.

Rejected Ideas
==============

``sys.implementation._simulator`` availability
----------------------------------------------

The ``_simulator`` attribute could be provided on *all* platforms, returning ``False``. However, the attribute has no use outside of an iOS context.

Open Issues
===========

On-device testing
-----------------

CI testing on simulators can be accommodated reasonably easily; however, on-device testing
is much harder.

> However, on device testing may not be necessary. As a data point - Apple's Xcode Cloud solution doesn't provide on-device testing. They rely on the fact that the API is consistent between device and simulator, and ARM64 simulator testing is sufficient to reveal CPU-specific issues.

Reviewer 3: Two howevers in a row, maybe rewrite one of them?

require production of officially distributed iOS artefacts for use by end-users.

If/when iOS is updated to Tier 2 or 1 support, there should be a process for
producing iOS distribution artefacts. This could be in the form of an "embedded
distribution" analogous to the Windows embedded distribution, or as a Cocoapod
or Swift Package that could be added to an Xcode project.

Console and interactive usage
-----------------------------

Distribution of a traditional CPython REPL or interactive "python.exe" should
not be considered a goal of this work.

Mobile devices (including iOS) do not provide a TTY-style console. They do not
provide ``stdin``, ``stdout`` or ``stderr``. iOS provides a system log, and it
is possible to install a redirection so that all ``stdout`` and ``stderr``
content is redirected to the system log; but there is no analog for
``stdin``.

In addition, iOS places restrictions on downloading additional code at runtime
(as this behavior would be functionally indistinguishable from trying to work
around App Store review). As a result, a traditional "create a virtual
environment and pip install" development experience will not be viable on iOS.

> It is *possible* to build an native iOS application the provides a REPL

Reviewer 3: ```suggestion
It is *possible* to build an native iOS application that provides a REPL
```

<https://cibuildwheel.readthedocs.io/>`__ may be one way to achieve this.

Security Implications
=====================

Adding iOS as a new platform does not add any security implications.

How to Teach This
=================

The education needs related to this PEP mostly relate to how end-users can add
iOS support to their own Xcode projects. This can be accomplished with
documentation and tutorials on that process. The need for this documentation
will increase if/when support raises from Tier 3 to Tier 2 or 1; however, this
transition should also be accompanied with simplified deployment artefacts (such
as a Cocoapod or Swift package) that are integrated with Xcode development.

Reference Implementation
========================

The BeeWare `Python-Apple-support
<https://github.com/beeware/Python-Apple-support>`__ repository contains a
reference patch and build tooling to compile a distributable artefact.

`Briefcase <https://briefcase.readthedocs.org>`__ provides a reference
implementation of code to execute test suites on iOS simulators. The `Toga
Testbed <https://github.com/beeware/toga/tree/main/testbed>`__ is an example of

> a test suite that is executed on the iOS simulator using Github Actions.

Reviewer 3: ```suggestion
a test suite that is executed on the iOS simulator using GitHub Actions.
```

Rejected Ideas
==============

``sys.implementation._simulator`` availability
----------------------------------------------

The ``_simulator`` attribute could be provided on *all* platforms, returning
``False``. However, the attribute has no use outside of an iOS context.

Open Issues
===========

On-device testing
-----------------

CI testing on simulators can be accommodated reasonably easily.
On-device testing is much harder, as availability of device farms that could be

> configured to provide Buildbots or Github Action runners is limited.

Reviewer 3: ```suggestion
configured to provide Buildbots or Github Actions runners is limited.
```
